src/BootAppManager.ts

```diff
@@ -1,26 +1,105 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT license.
 
-import { BootApp } from "./BootApp";
+import { BootApp, STATE_INACTIVE } from "./BootApp";
+
+import * as vscode from 'vscode';
+import * as uuid from 'uuid';
+import * as path from 'path';
+
+interface JavaProjectData {
+    path : string
+    name : string
+    classpath : ClassPathData
+}
+
+interface ClassPathData {
+    entries : CPE[];
+}
+
+interface CPE {
+    kind: string;
+    path: string; // TODO: Change to File, Path or URL?
+	outputFolder : string;
+	sourceContainerUrl : string;
+	javadocContainerUrl : string;
+	isSystem : boolean;
+}
+
+function isBootAppClasspath(cp : ClassPathData) : boolean {
+    if (cp.entries) {
+        let entries = cp.entries;
+        for (let i = 0; i < entries.length; i++) {
+            const cpe = entries[i];
+            let filename = path.basename(cpe.path);
+
+            if (filename.endsWith('.jar') && filename.startsWith('spring-boot')) {
+                return true;
+            }
+        }
+    } 
+    return false;
+}
 
 export class BootAppManager {
-    private _appList: BootApp[];
+
+    private _boot_projects : Map<String, JavaProjectData> = new Map();
 
     constructor() {
-        this._appList = [];
         this.initAppListSync();
-
     }
 
     public getAppList(): BootApp[] {
-        return this._appList;
+        let apps : BootApp[] = [];
+        this._boot_projects.forEach(p => {
+            apps.push(new BootApp(p.name, STATE_INACTIVE));
+        });
+        return apps;
     }
 
     private initAppListSync(): void {
-        // TODO: this is just a mock.
-        this._appList.push(
-            new BootApp("app 1", "running"),
-            new BootApp("app 2", "inactive")
-        );
+        //TODO: The code below will fail if jdt language server has not yet been started
+        //  How should we deal with that?
+        const callbackId = uuid.v4();
+
+        vscode.commands.registerCommand(callbackId, (...args) => {
+            let location : string = args[0];
+            let name : string = args[1];
+            let isDeleted : boolean = args[2];
+            if (isDeleted) {
+                this._boot_projects.delete(location);
+            } else {
+                let entries : ClassPathData = args[3];
+                if (isBootAppClasspath(entries)) {
+                    this._boot_projects.set(location, {
+                        path: location,
+                        name: name,
+                        classpath: entries
+                    });
+                } else {
+                    this._boot_projects.delete(location);
+                }
+            }
+        });
+
+        let tries = 0;
```

Comment:
I think it's more clear and easy to maintain to use async/await here.  And the outer function name `initAppListSync` should better be changed to `initAppListAsync`.

```typescript
this._registerClasspathListener(callbackId);
```

with this private helper:
```typescript
    private async _registerClasspathListener(callbackId: string) {
        const RETRY_TIMES = 20;
        const WAIT_IN_SECOND = 2;
        let tries = 0;
        while (tries < RETRY_TIMES) {
            try {
                await vscode.commands.executeCommand('java.execute.workspaceCommand', 'sts.java.addClasspathListener', callbackId);
                break;
            } catch (error) {
                tries++;
                sleep(WAIT_IN_SECOND * 1000);
            }
        }

        if (tries >= RETRY_TIMES) {
            throw new Error(`Fail to register classpath listener after ${RETRY_TIMES} retries.`);
        }
    }
```

Reply:
Here is the sleep function wrapping the `setTimeout`: 
```typescript
function sleep(milliseconds: number): Promise<void> {
    return new Promise<void>((resolve) => setTimeout(resolve, milliseconds));
}
```